### ISSUE TITLE
Tmx MapLoader crashes when using file properties and does not correctly load Text nodes

### ISSUE THREAD

**Issue by Quillraven** [libgdx/libgdx#7427]
Issue as disucced in #7291

Regarding the problem for text nodes: I only needed the `text` value in my game which I did not see anywhere in the LibGDX model classes. Therefore, I have written my own loader that simply calls `loadObject` and enhances the properties afterward:

```kotlin
class TiledLoader(fileHandleResolver: FileHandleResolver) : TmxMapLoader(fileHandleResolver) {

    // since LibGDX does not load the "text" information of a Text object in Tiled, we need to overrule
    // LibGDX's loadObject method and parse the text info ourselves and add it to the properties of the mapObject
    override fun loadObject(map: TiledMap, objects: MapObjects, element: XmlReader.Element, heightInPixels: Float) {
        super.loadObject(map, objects, element, heightInPixels)
        val lastObject = objects.last()
        element.getChildByName("text")?.text?.let { objTxt ->
            lastObject.properties.put("text", objTxt)
        }
    }
}
```

Text is just one of the fields of a text node in Tiled. There is also Font, Word Wrap and Alignment from what I can see:
![image](https://github.com/user-attachments/assets/24a683fa-5bb1-40ec-922e-5a3ab17f5536)

Would be of course beneficial to load all of them correctly.

----

The second issue is when using a file property in Tiled. That simply crashes LibGDX. Imo, the problematic method is in `BaseTmxMapLoader`:

```java
protected Object castProperty(String name, String value, String type) {
        if (type == null) {
            return value;
        } else if (type.equals("int")) {
            return Integer.valueOf(value);
        } else if (type.equals("float")) {
            return Float.valueOf(value);
        } else if (type.equals("bool")) {
            return Boolean.valueOf(value);
        } else if (type.equals("color")) {
            String opaqueColor = value.substring(3);
            String alpha = value.substring(1, 3);
            return Color.valueOf(opaqueColor + alpha);
        } else {
            throw new GdxRuntimeException("Wrong type given for property " + name + ", given : " + type + ", supported : string, bool, int, float, color");
        }
    }
```

"file" is not listed and therefore the GdxRuntimeException is thrown. Imo, file should be handled the same way as a string. No casting necessary. I don't know if a `FileHandle` makes sense. That might be difficult because Tiled stores the relative path to the map, if I remember correctly. That can be an arbitrary path outside the assets folder. I would leave that up to the user to handle it properly but at least give them the string.

**Comment by Quillraven** [comment]
I just noticed that "class" property type is also not supported. Same reason as for the file properties. An issue was however already created by a different user: #7432

**Comment by Quillraven** [comment]
I have the same opinion as you. Since objects are not rendered and need to be handled by the user, I think doing it the same way for text objects is fine. 
Also, you can select any font in Tiled with whatever size and bold/italic setting, etc.. How would the TmxLoader and the renderer know where to get those fonts. Theoretically you could provide this with the TmxMapLoader.Parameters or whatever they are called but I think that doesn't make sense.

I personally just needed the information of the text in my game and then I use the MapObject rectangle information + text information to create an entity in my ECS world. A system then takes care to render it the way I need it, which was similar to Tiled but with a slightly different font/color.


**In short:** handle text objects like normal objects. Don't render them in the renderer but create a MapObject instance with the correct properties of Tiled.

**Comment by BoBIsHere86** [comment]
I'll put together a PR at some point which adds the TextObject and fixes that File string bug. (Obviously it will still have to wait until the json loader is merged as I will need to duplicate the code in the json loader as well and then test)

But for our use case, before I continue I believe we both agree the whole point of this would be to just provide an object for the user with all of the possible attributes available. And the user may implement it as he wishes.

Keeping in line with how we handle out other objects I built a draft TextMapObject class for this purpose and I modified the hideous TiledMapObjectLoadingTest, to include 3 text fields along with the other objects. You can see below and the section of code to render what you see in the test.

![textObjTest](https://github.com/user-attachments/assets/0f67d37f-7181-44d7-bd52-6811979bd8c7)


```
else if(mapObject instanceof TextMapObject){
				 batch.begin();
				 TextMapObject textMapObject = (TextMapObject)mapObject;

				 int alignment;
				 String halign = textMapObject.getHorizontalAlign();

				 switch (halign.toLowerCase()) {
				 case "center":
					  alignment = Align.center;
					  break;
				 case "right":
					  alignment = Align.right;
					  break;
				 default:
					 // Default is 'left alignment, also there is no Align 'justify' like Tiled
					  alignment = Align.left;
					  break;
				 }

				 textMapObjectFont.setColor(textMapObject.getColor());
				 // The text rendering starts from the baseline, causing the text to appear below the specified Y-coordinate.
				 // To align the text with the top of the bounding box (as it appears in Tiled), we add textMapObject.getHeight() to the Y position.
				 textMapObjectFont.draw(batch, textMapObject.getText(), textMapObject.getX(), textMapObject.getY() + textMapObject.getHeight(),
					 textMapObject.getWidth(), alignment, textMapObject.isWrap());
				 batch.end();

				 shapeRenderer.setColor(Color.DARK_GRAY);
				 // Drawn display bounding box of TextMapObject
				 shapeRenderer.begin(ShapeRenderer.ShapeType.Line);
				 shapeRenderer.rect(textMapObject.getX(), textMapObject.getY(), textMapObject.getWidth(), textMapObject.getHeight());
				 shapeRenderer.end();

				 //reset back to blue
				 shapeRenderer.setColor(Color.BLUE);
			}
```

The differences in the test compared to Tiled stem from the different default fonts used and the window sizing.
Also, I don't often deal with drawing bitmap fonts but it seems like when we draw text its begins drawing below our coordinates. So when it came drawing I compensated by adding the bounding height to the y Position to offset, so it matched.

If there's anything else we should be concerned about or suggestions. Let me know. And eventually I'll throw up a PR.

**Comment by Quillraven** [comment]
I think that looks fine! Please open a PR so we can review it.

One remark: would be great if the alignment of the TextObject returns the proper values of LibGDX Align instead of strings. Can be added directly to the parsing I guess.

**Comment by BoBIsHere86** [comment]
> I think that looks fine! Please open a PR so we can review it.
> 
> One remark: would be great if the alignment of the TextObject returns the proper values of LibGDX Align instead of strings. Can be added directly to the parsing I guess.

It's perfect that you bring that up. I was on the fence about it. In tiled the halign can be "left, center, right or justify", and valign could be "top , center or bottom", right.

In the case of halign's "justify" what would the result be, do we just default left? That's why I originally didn't 1 to 1, store them as the Align values. 

Would we still want them stored as strings though. Could there still be a use for a user to access it? We could always just have  extra utility methods in there as well, to get as string/ or get the Align values directly.

I'll make the PR a bit later and it can be discussed further in there.

**Comment by BoBIsHere86** [comment]
This issue can now be closed as it was fixed by the original submitter of the ticket in his PR #7534 which has now been merged.
Good work!
